File: TP2-Kurtan/TP2-Kurtan/Tabuleiro.pl
:- dynamic board/1, key_picked/1, key_visible/1, game_over/1.

% Target positions
place_is_target(1, 3).
place_is_target(2, 4).
place_is_gate(3, 2).
place_is_key(1, 1).

key_picked(false).
key_visible(false).
game_over(false).

board([
    ['X', 'X', 'X', 'X', 'X', 'X','X', 'X', 'X', 'X'],
    ['X', ' ', ' ', '*', ' ', ' ',' ', 'X', 'X', 'X'],
    ['X', ' ', ' ', '@', '*', ' ',' ', 'X', 'X', 'X'],
    ['X', ' ', 'G', 'P', '@', ' ',' ', 'X', 'X', 'X'],
    ['X', ' ', ' ', ' ', ' ', ' ',' ', 'X', 'X', 'X'],
    ['X', 'X', 'X', 'X', 'X', 'X','X', 'X', 'X', 'X']
]).

print_board([]).
print_board([Row | Rest]) :-
    print_row(Row), nl,
    print_board(Rest).

print_row([]).
print_row([Cell | Rest]) :-
    write(Cell), write(' '),
    print_row(Rest).

find_player(Board, X, Y) :-
    nth0(X, Board, Row),
    nth0(Y, Row, 'P').

move(_Direction) :-
    game_over(true), !,
    write('Game over. Use restart_game to play again.'), nl.

move(Direction) :-
    board(Board),
    find_player(Board, X, Y),
    ( move_player(Board, Direction, X, Y, TempBoard) ->
        maybe_show_key(TempBoard, FinalBoard),
        retract(board(Board)),
        assert(board(FinalBoard)),
        print_board(FinalBoard),
        check_win(FinalBoard)
    ;   write('Invalid move.'), nl, false).

move_player(Board, up, X, Y, NewBoard) :-
    NewX is X - 1,
    handle_move(Board, X, Y, NewX, Y, NewBoard).
move_player(Board, down, X, Y, NewBoard) :-
    NewX is X + 1,
    handle_move(Board, X, Y, NewX, Y, NewBoard).
move_player(Board, left, X, Y, NewBoard) :-
    NewY is Y - 1,
    handle_move(Board, X, Y, X, NewY, NewBoard).
move_player(Board, right, X, Y, NewBoard) :-
    NewY is Y + 1,
    handle_move(Board, X, Y, X, NewY, NewBoard).

handle_move(Board, X, Y, NewX, NewY, NewBoard) :-
    is_within_bounds(Board, NewX, NewY),
    nth0(NewX, Board, Row),
    nth0(NewY, Row, TargetCell),
    (
        % Move to an empty cell
        TargetCell = ' ',
        update_board(Board, X, Y, NewX, NewY, NewBoard)
    ;
        % Move to a gate cell(only with a key)
        TargetCell = 'G',
        key_picked(true),
        replace(Board, X, Y, ' ', TempBoard),
        replace(TempBoard, NewX, NewY, '♛', NewBoard),
        write('You won!'), nl,
        retract(game_over(false)),
        assert(game_over(true)),
        write('Game over. Use restart_game to play again.'), nl
    ;
        % Move to target '*'
        TargetCell = '*',
        update_board(Board, X, Y, NewX, NewY, NewBoard)
    ;
        % Move to key 'K'
        TargetCell = 'K',
        key_visible(true),
        retract(key_picked(false)),
        assert(key_picked(true)),
        write('Key picked up! Gate can now be opened.'), nl,
        update_board(Board, X, Y, NewX, NewY, NewBoard)
    ;
    
        % Push box '@' or '$'
        (member(TargetCell, ['@', '$']),
         push_box(Board, X, Y, NewX, NewY, TempBoard, TargetCell),
         update_board(TempBoard, X, Y, NewX, NewY, NewBoard))

    ), !.


push_box(Board, X, Y, NewX, NewY, NewBoard, BoxType) :-
    NextX is NewX + (NewX - X),
    NextY is NewY + (NewY - Y),
    is_within_bounds(Board, NextX, NextY),
    nth0(NextX, Board, Row),
    nth0(NextY, Row, NextCell),
    (NextCell = ' '; NextCell = '*'),
    update_box(Board, NewX, NewY, NextX, NextY, TempBoard, BoxType),
    NewBoard = TempBoard.


update_board(Board, X, Y, NewX, NewY, NewBoard) :-
    cell_at(Board, X, Y, OldCell),
    (OldCell = 'P', cell_under(Board, X, Y, UnderOld), true ; UnderOld = ' '),
    (UnderOld = '*' -> replace(Board, X, Y, '*', TempBoard) ; replace(Board, X, Y, ' ', TempBoard)),
    replace(TempBoard, NewX, NewY, 'P', NewBoard).

cell_at(Board, X, Y, Cell) :-
    nth0(X, Board, Row),
    nth0(Y, Row, Cell).

cell_under(_, X, Y, '*') :-
    place_is_target(X, Y).
cell_under(_, _, _, ' ').

update_box(Board, X, Y, NewX, NewY, NewBoard, BoxType) :-
    nth0(NewX, Board, Row),
    nth0(NewY, Row, DestCell),
    (
        DestCell = '*' ->
            NewBox = '$'
        ;
            NewBox = '@'
    ),
    replace(Board, X, Y, ' ', TempBoard1),
    (BoxType = '$', place_is_target(X, Y) ->
        replace(TempBoard1, X, Y, '*', TempBoard2)
    ;
        TempBoard2 = TempBoard1
    ),
    replace(TempBoard2, NewX, NewY, NewBox, NewBoard).


check_key_pickup(Board) :-
    \+ (member(Row, Board), member('@', Row)).

% maybe_show_key_state(+Board, +KeyVisibleIn, -KeyVisibleOut, -NewBoard)
maybe_show_key_state(Board, false, true, NewBoard) :-
    \+ (member(Row, Board), member('@', Row)),  % all boxes placed
    write('[INFO] Key appeared in the room.'), nl,
    place_key(Board, NewBoard).  % place the key
maybe_show_key_state(Board, KeyVisible, KeyVisible, Board).

maybe_show_key(BoardIn, BoardOut) :-
    key_visible(Visible),
    maybe_show_key_state(BoardIn, Visible, NewVisible, TempBoard),
    (Visible \= NewVisible ->
        retractall(key_visible(_)),
        assert(key_visible(NewVisible))
    ;
        true
    ),
    BoardOut = TempBoard.

place_key(Board, NewBoard) :-
    place_key_in_empty(Board, NewBoard).

place_key_in_empty([], []).
place_key_in_empty([Row | Rest], [NewRow | Rest]) :-
    replace_first_empty(Row, NewRow),
    !.
place_key_in_empty([Row | Rest], [Row | NewRest]) :-
    place_key_in_empty(Rest, NewRest).

replace_first_empty([' ' | Rest], ['K' | Rest]) :- !.
replace_first_empty([Cell | Rest], [Cell | NewRest]) :-
    replace_first_empty(Rest, NewRest).


check_win(_Board) :-
    game_over(true),
    !.
check_win(_).

is_within_bounds(Board, X, Y) :-
    length(Board, NumRows),
    nth0(0, Board, Row),
    length(Row, NumCols),
    X >= 0, X < NumRows,
    Y >= 0, Y < NumCols.

replace([Row | Rest], 0, Y, NewElem, [NewRow | Rest]) :-
    replace_row(Row, Y, NewElem, NewRow).
replace([Row | Rest], X, Y, NewElem, [Row | NewRest]) :-
    X > 0,
    X1 is X - 1,
    replace(Rest, X1, Y, NewElem, NewRest).

replace_row([_ | Rest], 0, NewElem, [NewElem | Rest]).
replace_row([Cell | Rest], Y, NewElem, [Cell | NewRest]) :-
    Y > 0,
    Y1 is Y - 1,
    replace_row(Rest, Y1, NewElem, NewRest).

% Short commands
u :- move(up).
d :- move(down).
l :- move(left).
r :- move(right).

start_game :-
    retractall(board(_)),
    retractall(key_picked(_)),
    retractall(key_visible(_)),
    retractall(game_over(_)),
    assert(key_picked(false)),
    assert(key_visible(false)),
    assert(game_over(false)),
    Board = [
        ['X', 'X', 'X', 'X', 'X', 'X','X', 'X', 'X', 'X'],
        ['X', ' ', ' ', '*', ' ', ' ',' ', 'X', 'X', 'X'],
        ['X', ' ', ' ', '@', '*', ' ',' ', 'X', 'X', 'X'],
        ['X', ' ', 'G', 'P', '@', ' ',' ', 'X', 'X', 'X'],
        ['X', ' ', ' ', ' ', ' ', ' ',' ', 'X', 'X', 'X'],
        ['X', 'X', 'X', 'X', 'X', 'X','X', 'X', 'X', 'X']
    ],
    assert(board(Board)),
    print_board(Board).

restart_game :-
    write('Restarting the game...'), nl,
    start_game.

File: TP2-Kurtan/TP2-Kurtan/tabuleiro.py
from copy import deepcopy

WALL = 'X'
PLAYER = 'P'
BOX = '@'
BOX_ON_TARGET = '$'
TARGET = '*'
GATE = 'G'
KEY = 'K'
EMPTY = ' '

TARGET_POSITIONS = [(1, 3), (1, 5), (2, 4)]

class GameState:
    def __init__(self, board=None, key_picked=False, key_visible=False, game_over=False):
        self.board = board if board else self.initial_board()
        self.key_picked = key_picked
        self.key_visible = key_visible
        self.game_over = game_over
        self.player_pos = self.find_player()

    def initial_board(self):
        return [
            ['X', 'X', 'X', 'X', 'X', 'X','X', 'X', 'X', 'X'],
            ['X', 'X', ' ', '*', ' ', '*','X', 'X', 'X', 'X'],
            ['X', ' ', ' ', '@', '*', ' ','X', 'X', 'X', 'X'],
            ['X', ' ', 'G', 'P', '@', '@','X', 'X', 'X', 'X'],
            ['X', ' ', ' ', ' ', ' ', ' ',' ', 'X', 'X', 'X'],
            ['X', 'X', 'X', 'X', 'X', 'X','X', 'X', 'X', 'X']
        ]

    def find_player(self):
        for i, row in enumerate(self.board):
            for j, cell in enumerate(row):
                if cell == PLAYER:
                    return (i, j)
        return None

    def is_within_bounds(self, x, y):
        return 0 <= x < len(self.board) and 0 <= y < len(self.board[0])

    def cell_at(self, x, y):
        return self.board[x][y] if self.is_within_bounds(x, y) else None

    def cell_under(self, x, y):
        return TARGET if (x, y) in TARGET_POSITIONS else EMPTY

    def move(self, direction):
        if self.game_over:
            return False, "Game is over."

        dx, dy = direction_to_delta(direction)
        px, py = self.player_pos
        nx, ny = px + dx, py + dy

        if not self.is_within_bounds(nx, ny):
            return False, "Out of bounds."

        target_cell = self.cell_at(nx, ny)

        if target_cell in [EMPTY, TARGET]:
            self.update_board(px, py, nx, ny)
        elif target_cell == GATE and self.key_picked:
            self.update_board(px, py, nx, ny, win=True)
            self.game_over = True
        elif target_cell == KEY and self.key_visible:
            self.key_picked = True
            self.update_board(px, py, nx, ny)
        elif target_cell in [BOX, BOX_ON_TARGET]:
            if self.push_box(px, py, nx, ny, dx, dy, target_cell):
                self.update_board(px, py, nx, ny)
            else:
                return False, "Cannot push box."
        else:
            return False, "Invalid movement."

        if self.check_all_boxes_on_targets() and not self.key_visible:
            self.key_visible = True
            self.place_key()

        return True, "Moved."

    def update_board(self, px, py, nx, ny, win=False):
        if (px, py) in TARGET_POSITIONS:
            self.board[px][py] = TARGET
        else:
            self.board[px][py] = EMPTY

        self.board[nx][ny] = '♛' if win else PLAYER
        self.player_pos = (nx, ny)

    def push_box(self, px, py, bx, by, dx, dy, box_type):
        nx, ny = bx + dx, by + dy
        if not self.is_within_bounds(nx, ny):
            return False

        dest_cell = self.cell_at(nx, ny)
        if dest_cell not in [EMPTY, TARGET]:
            return False

        # move box
        self.board[nx][ny] = BOX_ON_TARGET if dest_cell == TARGET else BOX
        self.board[bx][by] = TARGET if (bx, by) in TARGET_POSITIONS else EMPTY
        return True

    def check_all_boxes_on_targets(self):
        for x, y in TARGET_POSITIONS:
            if self.board[x][y] not in [BOX_ON_TARGET]:
                return False
        return True

    def place_key(self):
        for i, row in enumerate(self.board):
            for j, cell in enumerate(row):
                if cell == EMPTY:
                    self.board[i][j] = KEY
                    return
                
    def is_goal_state(self):
        return self.check_all_boxes_on_targets() and self.key_picked and self.game_over

    def clone(self):
        new_state = GameState(deepcopy(self.board), self.key_picked, self.key_visible, self.game_over)
        new_state.player_pos = self.player_pos
        return new_state

    def print_board(self):
        for row in self.board:
            print(' '.join(row))
        print()

def direction_to_delta(direction):
    return {
        'up': (-1, 0),
        'down': (1, 0),
        'left': (0, -1),
        'right': (0, 1)
    }[direction]
